cmd/ceremony/main.go

```diff
@@ -380,169 +565,185 @@ func (cc crlConfig) validate() error {
 	}
+	template.AuthorityKeyId = issuer.SubjectKeyId
 
-	err = signAndWriteCert(template, template, keyInfo.key, signer, config.Outputs.CertificatePath, config.SkipLints)
+	lintCert, err := issueLintCert(template, issuer, pub, signer, config.SkipLints)
+	if err != nil {
+		return err
+	}
+
+	// Verify that the lintCert (and therefore the eventual finalCert) corresponds to the specified issuer certificate.
+	if !bytes.Equal(issuer.RawSubject, lintCert.RawIssuer) {
+		return fmt.Errorf("mismatch between issuer RawSubject and lintCert RawIssuer DER bytes: \"%x\" != \"%x\"", issuer.RawSubject, lintCert.RawIssuer)
+	}
+
+	_, err = signAndWriteCert(template, issuer, pub, signer, config.Outputs.CertificatePath)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func intermediateCeremony(configBytes []byte, ct certType) error {
-	var config intermediateConfig
+func crossCertCeremony(configBytes []byte, ct certType) error {
+	if ct != crossCert {
+		return fmt.Errorf("wrong certificate type provided")
+	}
+
+	var config crossCertConfig
 	err := strictyaml.Unmarshal(configBytes, &config)
 	if err != nil {
 		return fmt.Errorf("failed to parse config: %s", err)
 	}
-	err = config.validate(ct)
+	err = config.validate()
 	if err != nil {
 		return fmt.Errorf("failed to validate config: %s", err)
 	}
 
-	pubPEMBytes, err := os.ReadFile(config.Inputs.PublicKeyPath)
+	pub, pubBytes, err := loadPubKey(config.Inputs.PublicKeyPath)
 	if err != nil {
-		return fmt.Errorf("failed to read public key %q: %s", config.Inputs.PublicKeyPath, err)
-	}
-	pubPEM, _ := pem.Decode(pubPEMBytes)
-	if pubPEM == nil {
-		return fmt.Errorf("failed to parse public key")
-	}
-	pub, err := x509.ParsePKIXPublicKey(pubPEM.Bytes)
-	if err != nil {
-		return fmt.Errorf("failed to parse public key: %s", err)
+		return err
 	}
+
 	issuer, err := loadCert(config.Inputs.IssuerCertificatePath)
 	if err != nil {
 		return fmt.Errorf("failed to load issuer certificate %q: %s", config.Inputs.IssuerCertificatePath, err)
 	}
 
+	toBeCrossSigned, err := loadCert(config.Inputs.CertificateToCrossSignPath)
+	if err != nil {
+		return fmt.Errorf("failed to load toBeCrossSigned certificate %q: %s", config.Inputs.CertificateToCrossSignPath, err)
+	}
+
 	signer, randReader, err := openSigner(config.PKCS11, issuer.PublicKey)
 	if err != nil {
 		return err
 	}
 
-	template, err := makeTemplate(randReader, &config.CertProfile, pubPEM.Bytes, ct)
+	template, err := makeTemplate(randReader, &config.CertProfile, pubBytes, ct)
 	if err != nil {
 		return fmt.Errorf("failed to create certificate profile: %s", err)
 	}
 	template.AuthorityKeyId = issuer.SubjectKeyId
 
-	err = signAndWriteCert(template, issuer, pub, signer, config.Outputs.CertificatePath, config.SkipLints)
+	lintCert, err := issueLintCert(template, issuer, pub, signer, config.SkipLints)
 	if err != nil {
 		return err
 	}
 
+	// Ensure that we've configured the correct certificate to cross-sign compared to the profile.
```

Comment:
Other checks we can/should do:
- BRs 7.1.2.2.1: the cross-sign's notBefore date cannot be before the existing cert's notBefore
- BRs 7.1.2.2.3 to enable the next bullet point: we should enforce that the issuer and this cert have the same Organization Name in their subject, to show that this cross sign is allowed to be "unrestricted"
- BRs 7.1.2.2.3 plus some additional opinion: the cross-sign's EKU extension should match the existing cert's EKU extension (i.e. if the original is a Root with no EKUs, then this cross-sign can also have no EKUs; if the original is a SubCA with EKUs, this this cross-sign should also have those same EKUs)

Reply:
> BRs 7.1.2.2.3 to enable the next bullet point: we should enforce that the issuer and this cert have the same Organization Name in their subject, to show that this cross sign is allowed to be "unrestricted"

I don't think this is possible. I am doing a `bytes.Equal` check for `issuer.RawSubject` and `lintCert.RawIssuer` though which shows ".. the Cross-Certificate is operated by the same organization as the Issue CA ...".
```
[Internet Security Research Group] issuer.Issuer.OrganizationName
[Internet Security Research Group] issuer.Subject.OrganizationName

[Internet Security Research Group] lintCert.Issuer.OrganizationName
[(FAKE) Let's Encrypt]             lintCert.Subject.OrganizationName
```

Reply:
I think we want something like (pseudocode)
```go
if lintCert has EKUs {
  // we're in the "Restricted" case, so it's fine for organization names to be different.
  pass
} else {
  // we're in the "Unrestricted" case, so the issuer and subject need to be the same or at least affiliates. For us, we can go with just being the same.
  if lintCert.Subject.OrganizationName != issuer.Subject.OrganizationName {
    return fmt.Errorf("attempted unrestricted cross-sign of certificate operated by a different organization")
}
```

Reply:
@aarongable I think my latest commit addresses this.